lib/Sema/NameBinding.cpp

```diff
@@ -158,236 +380,417 @@ static const char *getImportKindString(ImportKind kind) {
   llvm_unreachable("Unhandled ImportKind in switch.");
 }
 
-static bool shouldImportSelfImportClang(const ImportDecl *ID,
-                                        const SourceFile &SF) {
+static bool shouldImportSelfImportClang(bool isExported, const SourceFile &SF) {
   // FIXME: We use '@_exported' for fake overlays in testing.
-  if (ID->isExported())
+  if (isExported)
     return true;
   if (SF.Kind == SourceFileKind::SIL)
     return true;
   return false;
 }
 
-void NameBinder::addImport(
-    SmallVectorImpl<SourceFile::ImportedModuleDesc> &imports, ImportDecl *ID) {
-  if (ID->getModulePath().front().Item == SF.getParentModule()->getName() &&
-      ID->getModulePath().size() == 1 && !shouldImportSelfImportClang(ID, SF)) {
-    // If the imported module name is the same as the current module,
-    // produce a diagnostic.
-    StringRef filename = llvm::sys::path::filename(SF.getFilename());
-    if (filename.empty())
-      Context.Diags.diagnose(ID, diag::sema_import_current_module,
-                             ID->getModulePath().front().Item);
-    else
-      Context.Diags.diagnose(ID, diag::sema_import_current_module_with_file,
-                             filename, ID->getModulePath().front().Item);
-    ID->setModule(SF.getParentModule());
-    return;
-  }
+bool UnboundImport::checkNotTautological(const SourceFile &SF) {
+  if (modulePath.front().Item != SF.getParentModule()->getName() ||
+      modulePath.size() != 1 ||
+      shouldImportSelfImportClang(options.contains(ImportFlags::Exported), SF))
+    return true;
 
-  ModuleDecl *M = getModule(ID->getModulePath());
-  if (!M) {
-    SmallString<64> modulePathStr;
-    interleave(ID->getModulePath(),
-               [&](ImportDecl::AccessPathElement elem) {
-                 modulePathStr += elem.Item.str();
-               },
-               [&] { modulePathStr += "."; });
-
-    auto diagKind = diag::sema_no_import;
-    if (SF.Kind == SourceFileKind::REPL || Context.LangOpts.DebuggerSupport)
-      diagKind = diag::sema_no_import_repl;
-    diagnose(ID->getLoc(), diagKind, modulePathStr);
-
-    if (Context.SearchPathOpts.SDKPath.empty() &&
-        llvm::Triple(llvm::sys::getProcessTriple()).isMacOSX()) {
-      diagnose(SourceLoc(), diag::sema_no_import_no_sdk);
-      diagnose(SourceLoc(), diag::sema_no_import_no_sdk_xcrun);
-    }
-    return;
-  }
+  ASTContext &Context = SF.getASTContext();
```

Comment:
Capitalizing this didn’t save any code later in the function from being removed and re-added.

Reply:
Changed.